swap/htlc.go

```diff
@@ -212,90 +296,219 @@ func swapHTLCScriptV1(cltvExpiry int32, senderHtlcKey,
 
 	builder.AddOp(txscript.OP_CHECKSIG)
 
-	return builder.Script()
-}
+	script, err := builder.Script()
+	if err != nil {
+		return nil, err
+	}
 
-// GenSuccessWitness returns the success script to spend this htlc with the
-// preimage.
-func (h *Htlc) GenSuccessWitness(receiverSig []byte,
-	preimage lntypes.Preimage) (wire.TxWitness, error) {
+	return &HtlcScriptV1{
+		script: script,
+	}, nil
+}
 
-	if h.Hash != preimage.Hash() {
-		return nil, errors.New("preimage doesn't match hash")
-	}
+// genSuccessWitness returns the success script to spend this htlc with
+// the preimage.
+func (h *HtlcScriptV1) genSuccessWitness(receiverSig []byte,
+	preimage lntypes.Preimage) wire.TxWitness {
 
 	witnessStack := make(wire.TxWitness, 3)
 	witnessStack[0] = append(receiverSig, byte(txscript.SigHashAll))
 	witnessStack[1] = preimage[:]
-	witnessStack[2] = h.Script
+	witnessStack[2] = h.script
 
-	return witnessStack, nil
+	return witnessStack
+}
+
+// GenTimeoutWitness returns the timeout script to spend this htlc after
+// timeout.
+func (h *HtlcScriptV1) GenTimeoutWitness(senderSig []byte) wire.TxWitness {
+
+	witnessStack := make(wire.TxWitness, 3)
+	witnessStack[0] = append(senderSig, byte(txscript.SigHashAll))
+	witnessStack[1] = []byte{0}
+	witnessStack[2] = h.script
+
+	return witnessStack
 }
 
 // IsSuccessWitness checks whether the given stack is valid for redeeming the
 // htlc.
-func (h *Htlc) IsSuccessWitness(witness wire.TxWitness) bool {
+func (h *HtlcScriptV1) IsSuccessWitness(witness wire.TxWitness) bool {
 	if len(witness) != 3 {
 		return false
 	}
 
 	isTimeoutTx := bytes.Equal([]byte{0}, witness[1])
+	return !isTimeoutTx
+}
+
+// Script returns the htlc script.
+func (h *HtlcScriptV1) Script() []byte {
+	return h.script
+}
+
+// MaxSuccessWitnessSize returns the maximum success witness size.
+func (h *HtlcScriptV1) MaxSuccessWitnessSize() int {
+	// Calculate maximum success witness size
+	//
+	// - number_of_witness_elements: 1 byte
+	// - receiver_sig_length: 1 byte
+	// - receiver_sig: 73 bytes
+	// - preimage_length: 1 byte
+	// - preimage: 32 bytes
+	// - witness_script_length: 1 byte
+	// - witness_script: len(script) bytes
+	return 1 + 1 + 73 + 1 + 32 + 1 + len(h.script)
+}
+
+// MaxTimeoutWitnessSize return the maximum timeout witness size.
+func (h *HtlcScriptV1) MaxTimeoutWitnessSize() int {
+	// Calculate maximum timeout witness size
+	//
+	// - number_of_witness_elements: 1 byte
+	// - sender_sig_length: 1 byte
+	// - sender_sig: 73 bytes
+	// - zero_length: 1 byte
+	// - zero: 1 byte
```

Comment:
note that in segwit encoding a 0 value is actually empty, and is implied by `zero_length` being set to 0. if i'm not mistaken this actually overestimates by 1 byte (assuming max size signature)

Reply:
Isn't that only true if we actually skip pushing the `[]byte{0}` to the stack?

Reply:
This is v1 and there is no change here right? Maybe you are referring to this issue: https://github.com/lightninglabs/nautilus/issues/173